src/back/spv/writer.rs

```diff
@@ -1617,18 +1690,118 @@ impl Writer {
                 let image_sample_result_type_id =
                     self.get_type_id(&ir_module.types, image_sample_result_type)?;
 
-                let main_instruction = match level {
-                    crate::SampleLevel::Auto => {
-                        super::instructions::instruction_image_sample_implicit_lod(
+                let depth_id = match depth_ref {
+                    Some(handle) => {
+                        let (expr_id, _) =
+                            self.write_expression(ir_module, ir_function, handle, block, function)?;
+                        Some(expr_id)
+                    }
+                    None => None,
+                };
+
+                let mut main_instruction = match level {
+                    crate::SampleLevel::Zero => {
+                        let mut inst = super::instructions::instruction_image_sample(
+                            image_sample_result_type_id,
+                            id,
+                            SampleLod::Explicit,
+                            sampled_image_id,
+                            coordinate_id,
+                            depth_id,
+                        );
+
+                        //TODO: cache this!
+                        let zero_id = self.generate_id();
+                        let zero_inner = crate::ConstantInner::Scalar {
+                            width: 4,
+                            value: crate::ScalarValue::Float(0.0),
+                        };
```

Comment:
Can't we use [`get_type_id()`](https://github.com/gfx-rs/naga/blob/master/src/back/spv/writer.rs#L215) for this? You should be able to request the scalar, even though it isn't in the IR. If it doesn't exist, it will add it to [`lookup_type`](https://github.com/gfx-rs/naga/blob/master/src/back/spv/writer.rs#L215).

Reply:
I know, I posted this already, and understand the difference. This mechanic is not implemented indeed for consts. I believe I have looked at it a couple of weeks/months ago, but there was a reason why I couldn't get it to work.

Reply:
Yeah, this is not a type ID, this is a constant. I think we can easily follow-up with the logic of caching these, but for the purposes of this PR, the correctness is more important than optimization.